### PR TITLE
Output ENR text representation in admin.nodeInfo RPC

### LINKED ISSUE #5622: Output ENR text representation in admin.nodeInfo RPC
See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-778.md#text-encoding for the relevant part of ENR spec.

1. Adapt base64 encoding/decoding functions to support URL-safe encoding (basically use `-` instead of `+`, `_` instead of `/` and don't pad with `=`)
https://github.com/ethereum/aleth/blob/master/libdevcore/Base64.h

2. Add method to `ENR` class returning string with text form of ENR (serilize to RLP and base64-encode it)

3. Add `enr` field in the output of `admin_net_nodeInfo` https://github.com/ethereum/aleth/blob/6890d9a8548008419dd4bc054ba7908eac3a933f/libweb3jsonrpc/AdminNet.cpp#L27
You'll need to extend `dev::p2p::NodeInfo` struct for that and fill it in the method `Host::nodeInfo()` with `Host::enr()`.

### PR DESCRIPTION
**Pull request by twinstar26**
Fixes #5622

### CODE REVIEW

**Comment by twinstar26** [comment]
Found a bug where they are using "==" as pads and I am removing only one of them, actually I should be removing both.

File: libp2p/ENR.h [hunk 1]
```diff
@@ -64,6 +68,7 @@ class ENR
     bytes content() const;
     size_t contentRlpListItemCount() const { return m_keyValuePairs.size() * 2 + 1; }
     void streamContent(RLPStream& _s) const;
+    Host* h;
```
Comment (gumb0) [inline libp2p/ENR.h]: This is not correct (`ENR` can belong to some peer, not to `Host`) and not needed, just use `streamRLP` of this object in `encodeEnr`

File: libdevcore/Base64.h [hunk 2]
```diff
@@ -35,6 +35,7 @@ namespace dev
 {
 
 std::string toBase64(bytesConstRef _in);
+std::string toBase64url(bytesConstRef _in);
```
Comment (gumb0) [inline libdevcore/Base64.h]: Call it `toBase64URLSafe`

File: libp2p/ENR.h [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 #pragma once
 
+#include <libdevcore/Base64.h>
```
Comment (gumb0) [inline libp2p/ENR.h]: Move this to `ENR.cpp`

File: libp2p/ENR.h [hunk 4]
```diff
@@ -56,6 +58,8 @@ class ENR
     ENR update(
         std::map<std::string, bytes> const& _keyValuePair, SignFunction const& _signFunction) const;
 
+    std::string encodeEnr();
```
Comment (gumb0) [inline libp2p/ENR.h]: Call it `textEncoding()`

Comment (gumb0) [reply libp2p/ENR.h]: Also make it `const` method

File: libp2p/ENR.h [hunk 5]
```diff
@@ -36,6 +37,7 @@ class ENR
     /// Create with given sign function
     ENR(uint64_t _seq, std::map<std::string, bytes> const& _keyValuePairs,
         SignFunction const& _signFunction);
+    ENR() = default;
```
Comment (gumb0) [inline libp2p/ENR.h]: You shouldn't need this.

File: libp2p/Host.h [hunk 6]
```diff
@@ -240,7 +241,12 @@ class Host: public Worker
     }
 
     /// Get the node information.
-    p2p::NodeInfo nodeInfo() const { return NodeInfo(id(), (networkConfig().publicIPAddress.empty() ? m_tcpPublic.address().to_string() : networkConfig().publicIPAddress), m_tcpPublic.port(), m_clientVersion); }
+    p2p::NodeInfo nodeInfo() const {
+        ENR e;
```
Comment (gumb0) [inline libp2p/Host.h]: You will return empty ENR this way, not useful to anyone.
Get it from `enr()` method

Comment (gumb0) [reply libp2p/Host.h]: Also use clang-format for you changes as described in https://github.com/ethereum/aleth/blob/master/CONTRIBUTING.md

Comment (twinstar26) [reply libp2p/Host.h]: Thanks for the reviews!!

**Comment by twinstar26** [comment]
> Instead of processing the result of `toBase64` it's better to create another similar function, then make both functions implemented in terms of the third internal function, taking `bytesConstRef _in` and the chars that vary as parameters.
> It could be for example `string toBase64(bytesConstRef _in, char const* _base64_chars, bool _pad)`
> (then define two variants of `base64_chars` array to pass there; `_pad` flags enables padding with `=`)

Oh some sort of wrapper functionality??

**Comment by gumb0** [comment]
There will be one internal function, and then two public function will wrap internal one, passing it different arguments.

**Comment by gumb0** [comment]
Replaced by https://github.com/ethereum/aleth/pull/5701

**Pull request closed without merging**